src/codewhisperer/service/inlineCompletionService.ts

```diff
@@ -0,0 +1,452 @@
+    }
+
+    public disposeInlineCompletion() {
+        this.inlineCompletionProviderDisposable?.dispose()
+        this.inlineCompletionProvider = undefined
+    }
+
+    async onEditorChange() {
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+    }
+
+    async onFocusChange() {
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+    }
+
+    async onCursorChange(e: vscode.TextEditorSelectionChangeEvent) {
+        if (e.kind !== 1 && vscode.window.activeTextEditor === e.textEditor) {
+            this.referenceProvider?.removeInlineReference()
+        }
+    }
+
+    async clearInlineCompletionStates(editor: vscode.TextEditor | undefined) {
+        await HoverConfigUtil.instance.restoreHoverConfig()
+        vsCodeState.isCodeWhispererEditing = false
+        this.referenceProvider?.removeInlineReference()
+        RecommendationHandler.instance.cancelPaginatedRequest()
+        RecommendationHandler.instance.reportUserDecisionOfCurrentRecommendation(editor, -1)
+        RecommendationHandler.instance.clearRecommendations()
+        this.disposeInlineCompletion()
+        this.setCodeWhispererStatusBarOk()
+        this.disposeCommandOverrides()
+        this.clearRejectionTimer()
+    }
+
+    setReferenceInlineProvider(provider: ReferenceInlineProvider) {
+        this.referenceProvider = provider
+    }
+
+    setExtContext(context: ExtContext) {
+        this.extContext = context
+    }
+
+    async getPaginatedRecommendation(
+        client: DefaultCodeWhispererClient,
+        editor: vscode.TextEditor,
+        triggerType: CodewhispererTriggerType,
+        config: ConfigurationEntry,
+        autoTriggerType?: CodewhispererAutomatedTriggerType
+    ) {
+        await this.clearInlineCompletionStates(editor)
+        this.setCodeWhispererStatusBarLoading()
+        await HoverConfigUtil.instance.overwriteHoverConfig()
+        this.registerCommandOverrides()
+        RecommendationHandler.instance.checkAndResetCancellationTokens()
+        this.documentUri = editor.document.uri
+        try {
+            let page = 0
+            while (page < this.maxPage) {
+                await RecommendationHandler.instance.getRecommendations(
+                    client,
+                    editor,
+                    triggerType,
+                    config,
+                    autoTriggerType,
+                    true,
+                    page
+                )
+
+                if (RecommendationHandler.instance.checkAndResetCancellationTokens()) {
+                    RecommendationHandler.instance.reportUserDecisionOfCurrentRecommendation(editor, -1)
+                    RecommendationHandler.instance.clearRecommendations()
+                    this.setCodeWhispererStatusBarOk()
+                    return
+                }
+
+                if (!this.isCurrentCompletionInProgress()) {
```

Comment:
What's the point of this? Are we silently skipping an update if the completion provider is providing results?
The current impl. is not async so `isCurrentCompletionInProgress` should always be false

Reply:
I should just rename this function to `isCurrentSuggestionVisible`. It avoids an update if a suggestion is already visible.

It won't always be false.

isCurrentCompletionInProgress(): boolean {
        return this.inlineCompletionProvider?.getActiveItemIndex !== undefined
    }